### DISCUSSION TITLE
Feedback loop fix?

### DISCUSSION THREAD

**Discussion by shibin2018** [sparcians/map#462] [General]
how I process the feedback?
for example, there are two modules A and B, A has outport bind to B, and A needs the feedback signal of B, when I make a outport from B to A, it will generate  DAG CYCLE .

**Comment by klingaard** [comment]
Hi @shibin2018, without seeing your code, I will assume that the DAG cycle is due to the framework believing data propagation from A -> B is 0 cycles and that the feedback signal from B -> A is also 0 cycles.   The framework makes the assumption that the feedback signal from B to A will cause A to send data to B _in the same cycle_, and that B will then _resend_ the feedback signal to A _again_ causing a potential infinite loop:
```
  B.feedback(A) -> A.send(B) -> B.feeedback(A) -> A.send(B) -> ...
```
Ways to fix this:
- Add a cycle to the feedback InPort in A
- Add a cycle to the InPort in B for the data
- Disable auto-precedence in A [using this function](https://sparcians.github.io/map/classsparta_1_1Unit.html#adc39dc12b225483f222fd6627c623f94) from sparta::Unit

If you disable auto-precedence the framework will not make the assumption I mentioned above.  Instead, the framework will now rely on the modeler to list the precedence rules between ports and events within the unit as well as between units.  However, you now have the risk of introducing a DAG _during simulation_ instead, so be mindful.

BTW, I have a small presentation on how this works [here](https://docs.google.com/presentation/d/e/2PACX-1vS1BWtVv0x3qXKQWAeECe2gsF9cMG3Zp2HnXJw52grCAcl21lv3a9pLW6J0lZ32e5DWdZkFyUMcE_AI/pub?start=false&loop=false&delayms=3000#slide=id.g7c850bbd55_1_408).

**Comment by shibin2018** [comment]
ok, I will try it, Thank you very much!

**Comment by shibin2018** [comment]
Hi, @klingaard May I ask another question?
what is the differenece between uniqueEvent and SingleCycleUniqueEvent. I can not find the distinction from the definition

**Comment by klingaard** [comment]
Both `SingleCycleUniqueEvent` and `UniqueEvent` share one main behavior: the modeler can schedule either of them multiple times for the same cycle, but the event will _only be fired once_ for that cycle.
```
    my_unique_event_.schedule();  // This schedules for now
    my_unique_event_.schedule();  // ignored
    my_unique_event_.schedule();  // ignored
    my_unique_event_.schedule();  // ignored
    my_unique_event_.schedule(1); // This schedules for NOW + 1
    my_unique_event_.schedule(1); // ignored
```
The _difference_ between the two is simply performance and one constraint.  `SingleCycleUniqueEvent` can _only be scheduled_ for either time **now** or for 1 cycle into the future.  This allows `SingleCycleUniqueEvent`'s design to be more efficient than `UniqueEvent`.  By contrast `UniqueEvent` can be scheduled for **any** time -- now, 1 cycle from now, or _N_ cycles from now, but the design requires a slight performance penalty in doing so.

I re-read the documentation for these types of events and indeed it's not clear on the differentiation.  I'll fix that.  Thanks!

**Reply by shibin2018** [reply]
ok, I see, Thanks!

**Comment by shibin2018** [comment]
Hi, @klingaard, I come again!
here is another question, how can I realize the function of valid-ready like digital circuit using the dataInport and dataoutport? 
I find that when dataoutport send message by dataoutport.send(), the dataInport can receive it unconditionally, can we receive the mesage depend on some conditions?
Thank you!

**Reply by furuame** [reply]
> I find that when dataoutport send message by dataoutport.send(), the dataInport can receive it unconditionally, can we receive the mesage depend on some conditions?

Either adding constraints on Sender's side, or adding filtering if-else on receiver's side. Or you can use `cancelIf` method when you find you have to cancel the transfer on the certain port.

**Reply by shibin2018** [reply]
but the link function with the dataInport will be called automaically when message send to the receiver. if I want to  use the message under some condition, i need first store the meaasge and then use it?

**Reply by furuame** [reply]
Yes I meant you can do the following things:
- Check the condition on the sender's side to determine whether to "send";
- Check the condition on the receiver's side (i.e. adding if-else clauses in your handler) to determine whether to return early
- Use DataOutPort's public method `cancelIf`, then the DataInPort will not be called if the receiving data match your conditions.

**Reply by klingaard** [reply]
Zen is correct on his methodologies if you do not have a credit system established between the two communicating blocks.  

Other methods of doing what you intend:
- Set up a signal ready port between the two units to indicate "can accept."  See the [Skeleton Pipeline](https://sparcians.github.io/map/skeleton_example.html) example
- Set up another set of Data[In,Out]Ports between your units that communicate the number of allowable transfers:
```
  UnitA:
      DataOutPort<Data> data_out;
      DataInPort<int>   credits_in;
      int num_credits_from_unit_B = 0;

  UnitB:
      DataInPort<Data>  data_in;
      DataOutPort<Data> credits_out;
      sparta::Buffer    data_received;
```

**Reply by shibin2018** [reply]
ok, I will try it, Thank you very much!

**Comment by shibin2018** [comment]
Hi, @klingaard , How can I set the precedence beteween event and DataInPort if they are both in Tick phase?, I find that the callback function of dataInport will be called before the event automatically when data in, I need the opposite calling order. I try use >> opertator but it will generate compile error.  And you mention that there is  auto-precedence in  sparta::Unit, what is the auto-precedence? Thank you for answering!

**Comment by shibin2018** [comment]
Hi, @klingaard , How can I set the precedence beteween event and DataInPort if they are both in Tick phase?, I find that the callback function of dataInport will be called before the event automatically when data in, I need the opposite calling order. I try use >> opertator but it will generate compile error.  And you mention that there is  auto-precedence in  sparta::Unit, what is the auto-precedence? Thank you for answering!

**Reply by furuame** [reply]
So, is it your desired behavior: You'd like the Event happens before the data being sent to the Port?

**Reply by shibin2018** [reply]
yes, same cycle  I'd like event happens before dataInport callback function

**Reply by furuame** [reply]
I think there is no direct way to do that. I'm not sure your scenario, but you can consider:
- Move the event to earlier phases, like Update or PortUpdate
- Assign 1-cycle delay to the Input Port, that would make Port handler be triggered in PortUpdate phase of Cycle T+1

**Reply by klingaard** [reply]
Zhen is correct that sparta does not support a precedence rule that ensures a given event occurs before data arrives on a given InPort.  Not sure of the use case here -- why would you need an event to be fired before data arrives on an InPort?  How are they related?  

Zhen's solutions (putting the Event on an [earlier phase](https://sparcians.github.io/map/namespacesparta.html#aeec80d237ace9f07634c3d24c9aed9a0)) should work.

I don't see any issue allowing this functionality -- just need to establish a precedence between an Event and any InPort that are within the same phase.  

Please open a GH issue for this support.

**Comment by shibin2018** [comment]
> Zhen is correct that sparta does not support a precedence rule that ensures a given event occurs before data arrives on a given InPort. Not sure of the use case here -- why would you need an event to be fired before data arrives on an InPort? How are they related?
> 
> Zhen's solutions (putting the Event on an [earlier phase](https://sparcians.github.io/map/namespacesparta.html#aeec80d237ace9f07634c3d24c9aed9a0)) should work.
> 
> I don't see any issue allowing this functionality -- just need to establish a precedence between an Event and any InPort that are within the same phase.
> 
> Please open a GH issue for this support.

ok
